### PR TITLE
revive: Fix dust & child contract calls

### PR DESCRIPTION
**Pull request by pgherveou**
When transferring to self we should just return early as it's a noop.
Not doing so cause bug in `transfer_with_dust`

as we do 
```
from.dust -= from.dust
to.dust += to.dust
```

We end up with a value of dust - planck (that we burnt from to create dust amount) on the account

fix https://github.com/paritytech/contract-issues/issues/211

### CODE REVIEW

**Review by 0xRVE — APPROVED** [review]

File: substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol [hunk 1]
```diff
@@ -0,0 +1,7 @@
+contract CallSelfWithDust {
+    function f() external payable {}
```
Comment (0xOmarA) [inline substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol:2]: To check, if `f` here contained some code, this code will be executed and the only thing that we have a `NOOP` for is just the transfers, is that correct?

File: substrate/frame/revive/src/exec.rs [hunk 2]
```diff
@@ -1521,6 +1521,10 @@ where
 		storage_meter: &mut storage::meter::GenericMeter<T, S>,
 		exec_config: &ExecConfig<T>,
 	) -> DispatchResult {
+		if from == to || value.is_zero() {
```
Comment (0xOmarA) [inline substrate/frame/revive/src/exec.rs]: I assume this fix applies to pallet-revive as a whole and therefore it would fix this issue for EVM and PolkaVM. Is that correct?

File: substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol (hunk 1 above)
Comment (xermicus) [reply substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol:2]: ~~Yes this breaks the EVM semantics.~~

**Review by xermicus — COMMENTED** [review]
As Omar pointed out, this PR seems to break EVM semantics? Naivly, I'd assume the proper thing to do is fix the dust logic?

File: substrate/frame/revive/src/exec.rs (hunk 2 above)
Comment (pgherveou) [reply substrate/frame/revive/src/exec.rs]: yes I guess we could check by whitelisting this test in your CI PR 

https://github.com/paritytech/polkadot-sdk/pull/10071

let me update the test here we can make it run against pvm as well

File: substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol (hunk 1 above)
Comment (pgherveou) [reply substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol:2]: > To check, if `f` here contained some code, this code will be executed and the only thing that we have a `NOOP` for is just the transfers, is that correct?

yes we short-circuit the logic of the transfer_with_dust fn when from = to or value = 0 the rest does not change

Comment (pgherveou) [reply substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol:2]: ah right we should always check that from.balance > amount is that what you are suggesting @xermicus

File: substrate/frame/revive/src/exec.rs [hunk 3]
```diff
@@ -1550,6 +1546,10 @@ where
 				to: &mut AccountInfo<T>,
 				dust: u32,
 			) -> DispatchResult {
+				if from == to || dust == 0 {
```
Comment (xermicus) [inline substrate/frame/revive/src/exec.rs]: Can dust be non-zero if the value itself is zero? There is already an early return that checks for the value be zero in the `transfer` function.

File: substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol (hunk 1 above)
Comment (xermicus) [reply substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol:2]: Nevermind. I saw that there is already a zero balance check in the transfer function and short-circuited that this is implemented in logic further up.

File: substrate/frame/revive/src/exec.rs (hunk 2 above)
Comment (xermicus) [reply substrate/frame/revive/src/exec.rs]: Yeah it applies to both backends. The PVM syscall handler isn't supposed to handle low level details like that.

File: substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol (hunk 1 above)
Comment (pgherveou) [reply substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol:2]: Current implementation is wrong i believe we should ensure that the account has enough balance to do the transfer even if to == from

Will patch it when i het back

Comment (xermicus) [reply substrate/frame/revive/fixtures/contracts/call_self_with_dust.sol:2]: Yeah the logic is quite complex with dust. Maybe needs some more tests for these corner cases.
